src/lib/private_components/layered-map-resources/alter_image.js

```diff
@@ -1,39 +1,166 @@
+    Promise.all([load_image(map_base64), load_image(colormap_base64)])
+        .then(function([map_image, colormap_image]) {
 
-            const map_image = textures[0].texture
+            const gl = canvas.getContext('webgl', {premultipliedAlpha: false })
+            gl.getExtension('OES_texture_float')
 
-            canvas.width = textures[0].width;
-            canvas.height = textures[0].height;
+            canvas.width = map_image.width
+            canvas.height = map_image.height
 
-            const colormap = textures[1].texture
+            // Clear canvas
+            gl.clearColor(0, 0, 0, 0)
+            gl.clear(gl.COLOR_BUFFER_BIT)
+
+            // Tell WebGL how to convert from clip space ([-1, 1] x [-1, 1]) back to pixel space ([0, w] x [0, h]):
+            gl.viewport(0, 0, gl.canvas.width, gl.canvas.height)
+
+            const program = createProgram(
+                gl,
+                createShader(gl, gl.VERTEX_SHADER, vertexShaderSource),
+                createShader(gl, gl.FRAGMENT_SHADER, hillshading ? fragmentShaderSourceWithHillshading : fragmentShaderSourceWithoutHillshading)
+            )
+            
+            gl.useProgram(program)
+
+            bindBuffer(gl, 'a_texCoord', [0, 0, 1, 0, 0, 1, 0, 1, 1, 0, 1, 1])
+            bindBuffer(gl, 'a_position', [
+                0, 0,
+                map_image.width, 0,
+                0, map_image.height,
+                0, map_image.height,
+                map_image.width, 0,
+                map_image.width, map_image.height
+            ])
+
+            bindTexture(gl, 0, 'u_image', map_image)
+            bindTexture(gl, 1, 'u_colormap_frame', colormap_image)
+
+            gl.uniform2f(
+                gl.getUniformLocation(program, 'u_resolution_vertex'),
+                gl.canvas.width,
+                gl.canvas.height
+            )
+
+            gl.uniform1f(
+                gl.getUniformLocation(program, 'u_colormap_length'),
+                colormap_image.width
+            )
 
-            const color_buffer = greyscale2colormap(reglObj, canvas, map_image, colormap, hillshading);
             if (hillshading){
-                direct_light(reglObj, canvas, map_image, color_buffer, false);
+                gl.uniform2f(
+                    gl.getUniformLocation(program, 'u_resolution_fragment'),
+                    gl.canvas.width,
+                    gl.canvas.height
+                )
+
+                const light_direction = [1, 1, 1]
+                const vectorLength = Math.sqrt(light_direction[0]**2 + light_direction[1]**2 + light_direction[2]**2)
+        
+                gl.uniform3f(
+                    gl.getUniformLocation(program, 'u_light_direction'),
+                    light_direction[0] / vectorLength,
+                    light_direction[1] / vectorLength,
+                    light_direction[2] / vectorLength
+                )
+
+                const elevationScale = 0.03
```

Comment:
Should this be a prop?

Reply:
Yes 💡  There are some properties related to hillshading which should probably be adjustable:
 - Light direction
 - Elevation scale
 - The normalization in the "dot product".
I'll make a separate issue out of it 👍